widgetastic_manageiq/__init__.py

```diff
@@ -2174,3 +2179,40 @@ class ProviderEntity(BaseEntity):
     quad_entity = ProviderQuadIconEntity
     list_entity = ProviderListEntity
     tile_entity = ProviderTileIconEntity
+
+
+class DynamicTable(VanillaTable):
+    """Extend the widget.Table class to implement row_add for dynamic tables with an 'Actions'
+    column.
+    In these tables, the top or bottom row can be clicked to add a new row, and when it is
+    clicked the row is replaced (top or bottom) with a row containing fillable widgets.
+
+    Args:
+        action_row: index of the action row, generally 0 or -1, defaults to 0
+
+        See Widgetastic.widget.Table for more arguments
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.action_row = kwargs.pop('action_row', 0)  # pull this off and pass the rest up
+        super(DynamicTable, self).__init__(*args, **kwargs)
+
+    def row_add(self):
+        """Use the action-cell column widget to add a row
+
+        Clicks on the row directly, not the action button
+        """
+        try:
+            self[self.action_row].click()
+            return self.action_row
```

Comment:
In the case action_row is negative, then the negative index is relative to the end of the table, which can move in this case. I recommend that in the case action_row is negative, then before doing the click convert the negative index into positive so it is absolute. Then access the row with this value and also return it.